### PR TITLE
AVM: Make `txn FirstValidTime` and `block` opcode available in logicsigs

### PR DESCRIPTION
**Pull request by jannotti**
Thread the ledger into the GroupContext so it can be used during LogicSig evaluation.

To limit the chance of stateless/stateful bugs, the Ledger is "narrowed" to a LedgerForSignature interface that only exposes the past block headers.

### CODE REVIEW

File: data/transactions/logic/eval.go [hunk 1]
```diff
@@ -610,6 +620,9 @@ func EvalContract(program []byte, gi int, aid basics.AppIndex, params *EvalParam
 	if params.Ledger == nil {
 		return false, nil, errors.New("no ledger in contract eval")
 	}
+	if params.SigLedger == nil {
+		params.SigLedger = params.Ledger
```
Comment (algochoi) [inline data/transactions/logic/eval.go:624]: Is ledger copying okay in terms of performance in LogicSigs? Also should something similar be done in `EvalSignature` to initialize the `SigLedger` there?

File: data/transactions/logic/evalStateful_test.go [hunk 2]
```diff
@@ -2537,17 +2538,17 @@ func TestLatestTimestamp(t *testing.T) {
...
 	// related to lv and the current round, so we set the fv,lv more
 	// realistically.
 	txn.FirstValid = l.round() - 10
 	txn.LastValid = l.round() + 10
 
 	// l.round() is 0xffffffff+5 = 4294967300 in test ledger
-	testApp(t, "int 4294967299; block BlkSeed; len; int 32; ==", ep) // current - 1
-	testApp(t, "int 4294967300; block BlkSeed; len; int 32; ==", ep,
-		"not available") // can't get current round's blockseed
+	testApp(t, "int 4294967300; block BlkSeed; len; int 32; ==", ep)
```
Comment (algorandskiy) [inline data/transactions/logic/evalStateful_test.go:2549]: this is confusing.
cow.Round() is always the current (latest +1). So the old test is valid and the new one is not.

File: data/transactions/logic/eval.go (hunk 1 above)
Comment (jannotti) [reply data/transactions/logic/eval.go:624]: However, a final thing to work out is whether real ledger and the sig ledger, as I've passed them in, agree on what `Round()` is. I think the SigLedger is giving the last committed round, and the Ledger is giving the one being worked on. So I have to track that detail down and ensure the opcodes work the same in sigs and apps.

File: ledger/internal/eval.go [hunk 3]
```diff
@@ -1483,6 +1483,7 @@ type evalTxValidator struct {
 	txcache          verify.VerifiedTransactionCache
 	block            bookkeeping.Block
 	verificationPool execpool.BacklogPool
+	ledger           logic.LedgerForLogic
```
Comment (jasonpaulos) [inline ledger/internal/eval.go:1486]: I don't see where this field is ever assigned

Comment (jannotti) [reply ledger/internal/eval.go:1486]: Agreed. What on earth.

File: data/transactions/logic/evalStateful_test.go (hunk 2 above)
Comment (algorandskiy) [reply data/transactions/logic/evalStateful_test.go:2549]: my understanding is we supply cow to contracts and ledger to logicsig. But the same `Round()` method returns the different values.
Maybe one should be Latest() and another Round()?

Comment (jannotti) [reply data/transactions/logic/evalStateful_test.go:2549]: I think I know what's going on now.  I added a Round() method to ledger.Ledger, so I could use it as a LedgerForSignature.  I should have used Latest()+1, but used Latest(). I need to get them doing the same thing, and then the test will go back to what it was.

File: ledger/internal/eval.go (hunk 3 above)
Comment (jannotti) [reply ledger/internal/eval.go]: It seems as though LogicSigs can be evaluated along two paths, and I've clearly not fixed the path that uses `evalTxValidator`. Before I fix it, I'd like to find or write failing tests.

File: data/transactions/logic/evalStateful_test.go (hunk 2 above)
Comment (jannotti) [reply data/transactions/logic/evalStateful_test.go]: This should be back to making sense.

File: data/transactions/logic/eval.go [hunk 4]
```diff
@@ -4848,7 +4861,7 @@ func (cx *EvalContext) availableRound(r uint64) (basics.Round, error) {
 	if firstAvail > cx.txn.Txn.LastValid || firstAvail == 0 { // early in chain's life
 		firstAvail = 1
 	}
-	current := cx.Ledger.Round()
+	current := cx.SigLedger.Round()
 	round := basics.Round(r)
 	if round < firstAvail || round >= current {
```
Comment (algorandskiy) [inline data/transactions/logic/eval.go:4866]: this is interesting, Round() is either current, or latest, depending on a context, and the condition `round >= current` is technically correct for both contexts but looks as a maintainability problem.

Comment (jannotti) [reply data/transactions/logic/eval.go:4866]: I've changed it so that regardless of how the SigLedger is supplied (either as a copy of Ledger, or on its own for LogicSig evaluation) Round() is Latest()+1. I think that is fairly good, although I remain worried about transactions added to a long queue that will be dropped asking for a header that doesn't exist yet.

File: ledger/internal/applications.go [hunk 5]
```diff
@@ -54,6 +54,14 @@ type cowForLogicLedger interface {
...
 }
 
+func (cs *roundCowState) AccountData(addr basics.Address) (ledgercore.AccountData, error) {
+	record, err := cs.Get(addr, true)
+	if err != nil {
+		return ledgercore.AccountData{}, err
+	}
+	return record, nil
+}
+
```
Comment (michaeldiamant) [inline ledger/internal/applications.go:64]: @jannotti Sanity checking:  `AccountData` appears to duplicate a pre-existing definition several lines below.  Can the PR's addition be removed?

```suggestion
```

File: ledger/internal/eval.go (hunk 3 above)
Comment (jannotti) [reply ledger/internal/eval.go]: @algorandskiy do we have tests evalTxValidator? It looks to me like it's use in txn validation, so I don't understand how my code is passing most tests, including e2e tests, without a way to get block headers.

File: data/transactions/logic/assembler.go [hunk 6]
```diff
@@ -2269,7 +2269,7 @@ func disassemble(dis *disassembleState, spec *OpSpec) (string, error) {
 	}
 	if strings.HasPrefix(spec.Name, "bytec_") {
 		b := spec.Name[len(spec.Name)-1] - byte('0')
-		if int(b) < len(dis.intc) {
+		if int(b) < len(dis.bytec) {
```
Comment (tzaffi) [inline data/transactions/logic/assembler.go:2272]: How is this change related to the current issue?

Comment (jannotti) [reply data/transactions/logic/assembler.go:2272]: It's not, but it's a dumb cut-n-paste bug I found while wandering around.

File: ledger/internal/applications.go (hunk 5 above)
Comment (jannotti) [reply ledger/internal/applications.go:64]: It's not a duplicate, since it adds the method to roundCowState, not LogicLedger. Nonetheless, it is extraneous for now. I added while pursing a path would have made *roundCowState into a LedgerForSignature.

I'll remove.

File: ledger/internal/eval.go (hunk 3 above)
Comment (jannotti) [reply ledger/internal/eval.go]: I suppose the only ones that would fail are ones with a `block` of `txn FirstValidTime` in them (since only they access SigLedger).  But we have an e2e test that does that, and it passes.

Comment (jannotti) [reply ledger/internal/eval.go]: I'm committing a check in `EvalSignature` for SigLedger == nil. I suppose we should have it anyway, but my goal is to see if maybe we do have a test that exercises this code, but it was not failing because it was not using the new opcodes, so it didn't complain about the nil SigLedger.

File: data/transactions/logic/eval.go (hunk 1 above)
Comment (algochoi) [reply data/transactions/logic/eval.go:624]: Thanks - I'll resolve as the `SigLedger` check in `EvalSignature` looks good to me

**Comment by jannotti** [comment]
I've removed `Round()` from SigLedger. The rule for what blocks are available is now simpler and depends only on the transaction, which seems nice for LogicSigs.

But the big mystery is how to get a test to fail for the clearly erroneous way the `evalTxValidator` is initialized in ledger/internal/eval.go's `Eval`

File: data/transactions/logic/eval.go [hunk 7]
```diff
@@ -4848,10 +4873,13 @@ func (cx *EvalContext) availableRound(r uint64) (basics.Round, error) {
 	if firstAvail > cx.txn.Txn.LastValid || firstAvail == 0 { // early in chain's life
 		firstAvail = 1
 	}
-	current := cx.Ledger.Round()
+	lastAvail := cx.txn.Txn.FirstValid - 1
+	if lastAvail > cx.txn.Txn.FirstValid { // txn had a 0 in FirstValid
+		lastAvail = 0 // So nothing will be available
+	}
```
Comment (algorandskiy) [inline data/transactions/logic/eval.go:4879]: I like it

File: ledger/internal/applications.go [hunk 8]
```diff
@@ -44,7 +44,7 @@ type cowForLogicLedger interface {
 	SetKey(addr basics.Address, aidx basics.AppIndex, global bool, key string, value basics.TealValue, accountIdx uint64) error
 	DelKey(addr basics.Address, aidx basics.AppIndex, global bool, key string, accountIdx uint64) error
 
-	round() basics.Round
+	Round() basics.Round
```
Comment (algorandskiy) [inline ledger/internal/applications.go:47]: looks like it should not be exported

**Review by algorandskiy — COMMENTED** [review]
LGTM. One comment about `Round` -> `round`.
I'll take another look on morning

File: ledger/internal/applications.go (hunk 8 above)
Comment (jannotti) [reply ledger/internal/applications.go:47]: Thanks, I agree, that was left over from another path.  I'll revert.

File: data/transactions/logic/evalStateful_test.go [hunk 9]
```diff
@@ -2537,19 +2538,23 @@ func TestLatestTimestamp(t *testing.T) {
...
 	txn.FirstValid = l.round() - 10
 	txn.LastValid = l.round() + 10
 
+	// Keep in mind that proto.MaxTxnLife is 1500 in the test proto
+
 	// l.round() is 0xffffffff+5 = 4294967300 in test ledger
-	testApp(t, "int 4294967299; block BlkSeed; len; int 32; ==", ep) // current - 1
+
+	// These first two tests show that current-1 is not available now, though a
+	// resonable extension is to allow such access for apps (not sigs).
```
Comment (algorandskiy) [inline data/transactions/logic/evalStateful_test.go:2552]: nit: resonable -> reasonable

**Review by algorandskiy — APPROVED** [review]
